src/box/lua/config/applier/console.lua

```diff
@@ -31,11 +33,19 @@ local function apply(config)
     local listen_uri = socket_file_to_listen_uri(socket_file)
     log.debug('console.apply: %s', listen_uri)
 
+    if old_socket ~= nil and old_socket:name() ~= socket_file then
```

Comment:
`<socket object>:name()` doesn't react on changing a current working directory.

```lua
local fio = require('fio')
local console = require('console')

local s = console.listen('unix/:./foo.control')
print(s:name())
fio.mkdir('bar')
box.cfg({work_dir = 'bar'})
print(s:name())
```

It prints:

```
unix/:./foo.control
unix/:./foo.control
```

I see the following scenarios, where it fails.

<details>
<summary>Existing socket is not closed, while it should</summary>

* **Action:** start an instance with `process.work_dir` = `w`.
* Default console socket is `<process.work_dir>/var/run/i-001/tarantool.control`.
* The `console.listen()` call is performed before changing CWD, so it uses the `w/var/run/i-001/tarantool.control` path.
* CWD is changed to `w`.
* **Action:** update config with `console.socket` = `w/var/run/{{ instance_name }}/tarantool.control`.
* **Action:** call `config:reload()`.
* It is expected that the old socket is closed and the new `w/w/<...>` one is created.
* `<existing socket>:name()` returns `unix/:./w/var/run/i-001/tarantool.control` that looks the same as the previous one, but it should be without `w/` taking into account the changed CWD.
* **Check:** the old socket is not closed, while is should be.
</details>

<details>
<summary>Existing socket is closed, while it shouldn't</summary>

* **Action:** start an instance.
* **Action:** update config with `console.socket` = `w/../var/run/{{ instance_name }}/tarantool.control`.
* **Action:** call `config:reload()`.
* It is expected that the socket is kept as is, without re-creation.
* Actually, the socket is closed, because the relative paths are different.
* **Check:** I guess that the effect of the socket re-creation leads to closing connections with currently connected clients.
</details>

Please, add test cases that would implement these scenarios.

I think that we should store an absolute path to perform the comparison.

Reply:
work_dir option is non-dynamic and config:reload() doesn't change the work_dir path.
Is this a reason not to watch this case?

Reply:
The console starts before the first box.cfg() call, which changes the working directory.

Reply:
I don't really understand how to implement these scripts because `box.cfg()` is implicitly called when creating or starting a `cluster` in `test/config-luatest/console_test.lua`, as a result an attempt to change the working directory results in an error `work_dir option is non-dynamic`.

Reply:
The config is applied in the following steps:

https://github.com/tarantool/tarantool/blob/cfd4bf4685de6ee2dced133519ce34f71e3bb71b/src/box/lua/config/init.lua#L146-L154

`console` applier is executed first, then `box_cfg`.

Let's consider the startup + reload scenario.

* A user starts tarantool with `--name` and `--config`
* `console.apply()` -- console starts to listen
* `box_cfg.apply()` -- CWD is changed
* A user reloads the configuration.
* `console.apply()` -- now it should compare a previous socket path that is relative to the previous CWD and the new (possibly unchanged) socket path that is relative to the new CWD
* `box_cfg.apply()`

The key point is that the first `console.apply()` is performed before changing the CWD, but the second one (triggered by `config:reload()`) after changing the CWD.